File: frontend/src/types/play.ts
export interface Play {
  id: string
  game_id: string
  play_number: number
  sequence: number
  quarter: number
  game_clock_seconds: number
  down: number | null
  yards_to_go: number | null
  yard_line_from_own: number | null
  score_home: number
  score_away: number
  play_type: string | null
  description: string | null
  posteam_abbr: string | null
  created_at: string
}

File: backend/app/schemas/play.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.shap import ShapFeature


class PlayRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    game_id: uuid.UUID
    play_number: int
    sequence: int
    quarter: int
    game_clock_seconds: int
    down: int | None
    yards_to_go: int | None
    yard_line_from_own: int | None
    score_home: int
    score_away: int
    play_type: str | None
    description: str | None
    posteam_abbr: str | None = None
    created_at: datetime


class PlayWpRead(PlayRead):
    """PlayRead extended with win-probability, SHAP data, and possession team."""
    home_wp: float
    away_wp: float
    top_shap: list[ShapFeature]
